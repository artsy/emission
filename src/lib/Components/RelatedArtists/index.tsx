import * as React from "react"
import { createFragmentContainer, graphql } from "react-relay"

import { LayoutChangeEvent, StyleSheet, TextStyle, View, ViewProperties, ViewStyle } from "react-native"

import SerifText from "../Text/Serif"
import RelatedArtist from "./RelatedArtist"

interface Props extends ViewProperties, RelayProps {}

interface State {
  columns: number
  imageSize: {
    width: number
    height: number
  }
}

class RelatedArtists extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      columns: 0,
      imageSize: {
        width: 1,
        height: 1,
      },
    }
  }

  layoutState(currentLayout): State {
    const width = currentLayout.width
    const isPad = width > 600
    const isPadHorizontal = width > 900

    const columnCount = isPad ? (isPadHorizontal ? 4 : 3) : 2
    const marginWidth = 20
    const totalMargins = marginWidth * (columnCount - 1)

    const imageWidth = (width - totalMargins) / columnCount
    const imageHeight = imageWidth / 1.5

    return {
      columns: columnCount,
      imageSize: {
        width: Math.floor(imageWidth),
        height: Math.floor(imageHeight),
      },
    }
  }

  onLayout = (event: LayoutChangeEvent) => {
    const newLayoutState = this.layoutState(event.nativeEvent.layout)
    if (this.state.columns !== newLayoutState.columns) {
      this.setState(newLayoutState)
    }
  }

  render() {
    return (
      <View style={styles.container} onLayout={this.onLayout.bind(this)}>
        <SerifText style={styles.heading}>Related Artists</SerifText>
        <View style={styles.artistContainer}>
          {this.renderArtists()}
        </View>
      </View>
    )
  }

  renderArtists() {
    const artists = this.props.artists
    const artistViews = artists.map(artist => {
      return <RelatedArtist key={artist.__id} artist={artist} imageSize={this.state.imageSize} />
    })

    const numberOfTrailingViews = artists.length % this.state.columns
    if (numberOfTrailingViews > 0) {
      const extraRequiredViews = this.state.columns - numberOfTrailingViews
      for (let i = 0; i < extraRequiredViews; i++) {
        artistViews.push(<View key={"related-artist-spacer-" + i} style={this.state.imageSize} />)
      }
    }

    return artistViews
  }
}

interface Styles {
  container: ViewStyle
  artistContainer: ViewStyle
  heading: TextStyle
}

const styles = StyleSheet.create<Styles>({
  container: {
    flexDirection: "column",
  },
  artistContainer: {
    flexWrap: "wrap",
    flexDirection: "row",
    alignItems: "flex-start",
    justifyContent: "space-around",
    marginTop: 12,
    marginLeft: -10,
    marginRight: -10,
  },
  heading: {
    fontSize: 20,
  },
})

export default createFragmentContainer(
  RelatedArtists,
  graphql`
    fragment RelatedArtists_artists on Artist @relay(plural: true) {
      __id
      ...RelatedArtist_artist
    }
  `
)

interface RelayProps {
  artists: Array<{
    __id: string
  } | null> | null
}
