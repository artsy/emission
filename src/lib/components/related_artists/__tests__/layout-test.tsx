import 'react-native'
import * as React from 'react'
import { renderWithLayout } from '../../../tests/render_with_layout'

import RelatedArtists from '../'

it('lays out correctly', () => {
  const artists = [ {
    __id: 0,
    name: 'Sarah Scott',
    counts: {
      for_sale_artworks: 2,
      artworks: 4,
    },
  }]

  const layout = { width: 768 }

  const component = renderWithLayout(<RelatedArtists artists={artists}/>, layout)
  expect(component).toMatchSnapshot()
})
