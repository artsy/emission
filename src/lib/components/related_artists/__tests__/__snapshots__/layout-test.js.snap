exports[`test lays out correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flexDirection": "column",
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontSize": 17,
        },
        Object {
          "fontSize": 20,
        },
        Object {
          "fontFamily": "AGaramondPro-Regular",
        },
      ]
    }>
    Related Artists
  </Text>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-around",
        "marginLeft": -10,
        "marginRight": -10,
        "marginTop": 12,
      }
    }>
    <TouchableWithoutFeedback
      onPress={[Function]}>
      <View
        style={
          Object {
            "margin": 5,
            "paddingBottom": 20,
            "width": 242,
          }
        }>
        <ImageView
          imageURL={undefined}
          style={
            Object {
              "height": 161,
              "width": 242,
            }
          } />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontFamily": "Avant Garde Gothic ITCW01Dm",
              "fontSize": 12,
              "marginTop": 10,
              "textAlign": "left",
            }
          }>
          SARAH SCOTT
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#666666",
              "fontFamily": "AGaramondPro-Regular",
              "fontSize": 16,
              "marginTop": 5,
            }
          }>
          4 works, 2 for sale
        </Text>
      </View>
    </TouchableWithoutFeedback>
    <View
      style={
        Object {
          "height": 161,
          "width": 242,
        }
      } />
    <View
      style={
        Object {
          "height": 161,
          "width": 242,
        }
      } />
  </View>
</View>
`;
